core/src/main/java/org/infinispan/interceptors/locking/PessimisticLockingInterceptor.java

```diff
@@ -190,95 +145,89 @@ public Object visitLockControlCommand(TxInvocationContext ctx, LockControlComman
-         needRemoteLock = !localTransaction.getAffectedKeys().containsAll(keys);
-         if (!needRemoteLock) {
-            if (trace) log.tracef("We already have lock for keys %s, skip remote lock acquisition", keys);
-         }
-      }
-      return needRemoteLock;
+   private boolean skipLockingForKey(FlagAffectedCommand command, TxInvocationContext<?> ctx, Object key) {
+      return !ctx.isOriginLocal() ||
+             hasSkipLocking(command) ||
+             isKeysAlreadyLocked(ctx, key);
    }
 
-   private boolean needRemoteLocks(InvocationContext ctx, Object key, FlagAffectedCommand command)
-         throws Throwable {
-      boolean needBackupLocks = ctx.isOriginLocal() && (!isLockOwner(key) || isStateTransferInProgress());
-      boolean needRemoteLock = false;
-      if (needBackupLocks && !command.hasAnyFlag(FlagBitSets.CACHE_MODE_LOCAL)) {
-         final TxInvocationContext txContext = (TxInvocationContext) ctx;
-         LocalTransaction localTransaction = (LocalTransaction) txContext.getCacheTransaction();
-         needRemoteLock = !localTransaction.getAffectedKeys().contains(key);
-         if (!needRemoteLock) {
-            if (trace)
-               log.tracef("We already have lock for key %s, skip remote lock acquisition", key);
-         }
-      } else {
-         if (trace)
-            log.tracef("Don't need backup locks %s", needBackupLocks);
-      }
-      return needRemoteLock;
+   private boolean skipLockingForKeys(FlagAffectedCommand command, TxInvocationContext<?> ctx,
+         Collection<?> keys) {
+      return !ctx.isOriginLocal() ||
+             hasSkipLocking(command) ||
+             areAllKeysAlreadyLocked(ctx, keys);
    }
 
-   private boolean isLockOwner(Collection<?> keys) {
-      for (Object key : keys) {
-         if (!isLockOwner(key)) {
-            return false;
-         }
-      }
-      return true;
+   private boolean isLocalOnly(FlagAffectedCommand command) {
+      return command.hasAnyFlag(FlagBitSets.CACHE_MODE_LOCAL);
    }
 
-   private boolean isStateTransferInProgress() {
-      return stateTransferManager != null && stateTransferManager.isStateTransferInProgress();
+   private Object invokeLockCommandAndContinue(Collection<?> keys, FlagAffectedCommand command,
+         TxInvocationContext<?> ctx, int topologyId) throws InterruptedException {
+      //we need to acquire locks
+      LockControlCommand lcc = cf.buildLockControlCommand(keys, command.getFlagsBitSet(), ctx.getGlobalTransaction());
+      lcc.setTopologyId(topologyId);
+      return makeStage(handleLocalLockCommand(ctx, lcc))
+            .thenApply(ctx, command, (rCtx, rCommand, rv) -> invokeNext(rCtx, rCommand));
    }
+
+   private Object handleLocalLockCommand(TxInvocationContext context, LockControlCommand command)
+         throws InterruptedException {
+      //TODO blocking method. change it to non-blocking? [minor, only blocks local commands]
```

Comment:
Yes, this is unacceptable because you would be blocking `xxxAsync` calls!

Reply:
created a jira: https://issues.jboss.org/browse/ISPN-8336